Tests/SonarScanner.MSBuild.PreProcessor.Test/JreCaching/TarGzUnpackTests.cs

```diff
@@ -0,0 +1,140 @@
+﻿/*
+ * SonarScanner for .NET
+ * Copyright (C) 2016-2024 SonarSource SA
+ * mailto: info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.IO;
+using FluentAssertions;
+using ICSharpCode.SharpZipLib.Core;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using NSubstitute;
+using SonarScanner.MSBuild.Common;
+using SonarScanner.MSBuild.PreProcessor.JreCaching;
+using TestUtilities;
+
+namespace SonarScanner.MSBuild.PreProcessor.Test.JreCaching;
+
+[TestClass]
+public class TarGzUnpackTests
+{
+    [TestMethod]
+    public void TarGzUnpacking_Success()
+    {
+        // A tarball with the following content:
+        // Main
+        //  ├── Sub
+        //  └── Sub2
+        //      └── Sample.txt
+        const string sampleTarGzFile = """
+            H4sICL04jWYEAE1haW4udGFyAO3SUQrDIAyA4RzFE2wao55iTz2BBccK3Ribw
+            nb7iVDKnkqh+mK+l4S8/rn46XGGumTmnMuz+JvLrsiSRk1ImO/WkgRhoIH0jv
+            4lBHSq9B/SWPMHdvVHk+9OO+T+LSz9seID7OpPpT8Zy/1bWPsP/v6cwyl+Ihx
+            ss78ya3+T70qR1iAkNNB5/1v4ijH4FKdrmoExxlgvfmqGu7oADgAA
+            """;
+        var baseDirectory = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
+        var main = Path.Combine(baseDirectory, "Main");
+        var sub1 = Path.Combine(baseDirectory, "Main", "Sub");
+        var sub2 = Path.Combine(baseDirectory, "Main", "Sub2");
+        var sampleTxt = Path.Combine(baseDirectory, "Main", "Sub2", "Sample.txt");
+        var osProvider = Substitute.For<IOperatingSystemProvider>();
+        osProvider.OperatingSystem().Returns(PlatformOS.MacOSX);
+        using var archive = new MemoryStream(Convert.FromBase64String(sampleTarGzFile));
+        var sut = new TarGzUnpacker(DirectoryWrapper.Instance, FileWrapper.Instance, osProvider);
```

Comment:
Isn't the whole idea of the substitutes that you don't need to do actual I/O? I would use a substitute for the directory and the file and check what was received.  See e.g. how my tests looked like before I changed the zip implementation:
https://github.com/SonarSource/sonar-scanner-msbuild/blob/f269a83ab09016ded418f5d21ffd5f6bb28d5030/Tests/SonarScanner.MSBuild.PreProcessor.Test/JreCaching/ZipUnpackTests.cs

Before doing the change, we need to decide if we want to do what I proposed as an alternative approach here:
https://github.com/SonarSource/sonar-scanner-msbuild/pull/2036#discussion_r1673519191

Reply:
I think it is fine to copy the code as you suggested.
About the substitutes, I will be able to rewrite the tests, but the tar entries are extracted using this:

```csharp
tar.CopyEntryContents(outputStream);
```

...so if I do it with substitutes, I will not be able to assert:

```csharp
content.Should().Be("whatever");
```

If you think this is not that important, I can refactor it.
Please let me know :)

Reply:
...that's awesome.